File: web/src/views/partials/health.ts
import type { ClusterState } from "../../gen/types_pb.ts";

export interface HealthIssue {
  color: "red" | "yellow";
  label: string;
  count: number;
  link: string;
}

export function computeHealthIssues(state: ClusterState): HealthIssue[] {
  const issues: HealthIssue[] = [];

  // Stuck instances: assigned but not ready > 60s
  let stuckCount = 0;
  for (const sup of state.supervisors) {
    for (const inst of sup.instances) {
      if (inst.readyAt == null && inst.assignedAt != null) {
        const assignedMs = Number(inst.assignedAt.seconds) * 1000;
        if (Date.now() - assignedMs > 60_000) {
          stuckCount++;
        }
      }
    }
  }
  if (stuckCount > 0) {
    issues.push({ color: "red", label: "Stuck instances", count: stuckCount, link: "/functions" });
  }

  // Functions waiting for pods: has instances but none ready
  let waitingCount = 0;
  for (const sup of state.supervisors) {
    if (sup.instances.length > 0 && sup.instances.every((i) => i.readyAt == null)) {
      waitingCount++;
    }
  }
  if (waitingCount > 0) {
    issues.push({
      color: "yellow",
      label: "Functions waiting for pods",
      count: waitingCount,
      link: "/functions",
    });
  }

  // Stale heartbeats: > 60s old (approaching 90s timeout)
  let staleHeartbeatCount = 0;
  for (const sup of state.supervisors) {
    for (const hb of sup.routerHeartbeats) {
      if (hb.heartbeat?.timestamp != null) {
        const hbMs = Number(hb.heartbeat.timestamp.seconds) * 1000;
        if (Date.now() - hbMs > 60_000) {
          staleHeartbeatCount++;
        }
      }
    }
  }
  if (staleHeartbeatCount > 0) {
    issues.push({
      color: "yellow",
      label: "Stale heartbeats",
      count: staleHeartbeatCount,
      link: "/routers",
    });
  }

  // Stale instances: on non-active replica sets
  let staleInstanceCount = 0;
  for (const sup of state.supervisors) {
    if (sup.activeReplicaSet) {
      for (const inst of sup.instances) {
        if (inst.replicaSet !== sup.activeReplicaSet) {
          staleInstanceCount++;
        }
      }
    }
  }
  if (staleInstanceCount > 0) {
    issues.push({
      color: "yellow",
      label: "Stale instances",
      count: staleInstanceCount,
      link: "/functions",
    });
  }

  return issues;
}

export function healthIndicators(state: ClusterState): string {
  const issues = computeHealthIssues(state);

  if (issues.length === 0) {
    return `<div class="health-indicators">
      <div class="health-indicator">
        <div class="health-dot health-dot-green"></div>
        <span>All healthy</span>
      </div>
    </div>`;
  }

  const indicators = issues
    .map(
      (
        issue,
      ) => `<a href="${issue.link}" class="health-indicator" style="text-decoration: none; color: inherit">
      <div class="health-dot health-dot-${issue.color}"></div>
      <span>${issue.label}</span>
      <span class="badge badge-${issue.color}">${issue.count}</span>
    </a>`,
    )
    .join("\n");

  return `<div class="health-indicators">${indicators}</div>`;
}

export function healthPartial(state: ClusterState): Response {
  return new Response(healthIndicators(state), {
    headers: { "content-type": "text/html; charset=utf-8" },
  });
}

File: internal/web/health.go
package web

import (
	"time"

	"github.com/gadget-inc/skipper/internal/skipper"
)

type healthIssue struct {
	Color string
	Label string
	Count int
	Link  string
}

func computeHealthIssues(state *skipper.ClusterState) []healthIssue {
	var issues []healthIssue
	now := time.Now()

	// Stuck instances: assigned but not ready > 60s
	var stuckCount int
	for _, sup := range state.GetSupervisors() {
		for _, inst := range sup.GetInstances() {
			if !inst.HasReadyAt() && inst.HasAssignedAt() {
				if now.Sub(inst.GetAssignedAt().AsTime()) > 60*time.Second {
					stuckCount++
				}
			}
		}
	}
	if stuckCount > 0 {
		issues = append(issues, healthIssue{Color: "red", Label: "Stuck instances", Count: stuckCount, Link: "/functions"})
	}

	// Functions waiting for pods: has instances but none ready
	var waitingCount int
	for _, sup := range state.GetSupervisors() {
		instances := sup.GetInstances()
		if len(instances) > 0 {
			allPending := true
			for _, inst := range instances {
				if inst.HasReadyAt() {
					allPending = false
					break
				}
			}
			if allPending {
				waitingCount++
			}
		}
	}
	if waitingCount > 0 {
		issues = append(issues, healthIssue{Color: "yellow", Label: "Functions waiting for pods", Count: waitingCount, Link: "/functions"})
	}

	// Stale heartbeats: > 60s old
	var staleHeartbeatCount int
	for _, sup := range state.GetSupervisors() {
		for _, hb := range sup.GetRouterHeartbeats() {
			if hb.GetHeartbeat() != nil && hb.GetHeartbeat().HasTimestamp() {
				if now.Sub(hb.GetHeartbeat().GetTimestamp().AsTime()) > 60*time.Second {
					staleHeartbeatCount++
				}
			}
		}
	}
	if staleHeartbeatCount > 0 {
		issues = append(issues, healthIssue{Color: "yellow", Label: "Stale heartbeats", Count: staleHeartbeatCount, Link: "/routers"})
	}

	// Stale instances: on non-active replica sets
	var staleInstanceCount int
	for _, sup := range state.GetSupervisors() {
		activeRS := sup.GetActiveReplicaSet()
		if activeRS != "" {
			for _, inst := range sup.GetInstances() {
				if inst.GetReplicaSet() != activeRS {
					staleInstanceCount++
				}
			}
		}
	}
	if staleInstanceCount > 0 {
		issues = append(issues, healthIssue{Color: "yellow", Label: "Stale instances", Count: staleInstanceCount, Link: "/functions"})
	}

	return issues
}
